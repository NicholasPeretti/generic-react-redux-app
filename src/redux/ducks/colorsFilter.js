import { Map } from 'immutable'

export const NAMESPACE = 'COLOR_FILTER'
export const SET_COLORS = `${NAMESPACE} SET_COLORS`
export const SELECT_COLOR = `${NAMESPACE} SELECT_COLOR`

export const setColors = colors => ({
  type: SET_COLORS,
  colors
})

export const selectColor = color => ({
  type: SELECT_COLOR,
  color
})

export const defaultState = {
  selectedColor: null,
  colors: [],
  colorsMap: new Map()
}

export default function reducer (state = defaultState, action = {}) {
  switch (action.type) {
    case SET_COLORS: {
      state.colorsMap = new Map()
      action.colors.forEach(color => {
        state.colorsMap = state.colorsMap.set(
          color, color
        )
      })

      state = Object.assign({}, state, {
        colors: state.colorsMap.reduce((colors, v) => colors.concat([v]), [])
      })
      break
    }

    case SELECT_COLOR: {
      state = Object.assign({}, state, {
        selectedColor: action.color
      })
      break
    }
  }

  return state
}
