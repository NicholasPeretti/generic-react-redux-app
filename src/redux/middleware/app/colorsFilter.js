import {
  apiRequest,
  isApiAction,
  API_SUCCESS,
  API_ERROR,
  API_REQUEST
} from '../core/api'
import { GET_COLORS } from '../../endpoints'
import { NAMESPACE, setColors, setError, setFetching } from '../../ducks/colorsFilter'

export const fetchColors = () => {
  const { url, method } = GET_COLORS
  return apiRequest(url, method, NAMESPACE)
}

export default function middleware () {
  return dispatch => action => {
    dispatch(action)

    handleApiMiddlewareActions(dispatch, action)
  }
}

function handleApiMiddlewareActions (dispatch, action) {
  if (!isApiAction(action, NAMESPACE)) return

  switch (action.type) {
    case API_REQUEST: {
      dispatch(setFetching(true))
      dispatch(setError(null))
      break
    }

    case API_SUCCESS: {
      const { colors } = action.payload
      dispatch(setColors(colors))
      dispatch(setFetching(false))
      break
    }

    case API_ERROR: {
      const error = action.payload
      dispatch(setError(error))
      dispatch(setFetching(false))
      break
    }
  }
}
