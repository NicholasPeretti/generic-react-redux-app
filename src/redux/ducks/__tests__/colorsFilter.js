import DuckTest from '../../../utils/tests/duck'
import * as colorsFilter from '../colorsFilter'
import { Colors as MOCK_COLORS } from '../../../utils/tests/data'

describe('Colors Filter', () => {
  DuckTest(colorsFilter)

  test('defaultState should match snapshot', () => {
    expect(colorsFilter.defaultState).toMatchSnapshot()
  })

  describe('Action creators', () => {
    describe('setColors', () => {
      test('Should have the correct type', () => {
        const action = colorsFilter.setColors()
        expect(action.type).toEqual(colorsFilter.SET_COLORS)
      })

      test('Should have the correct value', () => {
        const action = colorsFilter.setColors(MOCK_COLORS)
        expect(action.colors).toEqual(MOCK_COLORS)
      })
    })

    describe('selectColor', () => {
      test('Should have the correct type', () => {
        const action = colorsFilter.selectColor()
        expect(action.type).toEqual(colorsFilter.SELECT_COLOR)
      })

      test('Should have the correct value', () => {
        const color = MOCK_COLORS[0]
        const action = colorsFilter.selectColor(color)
        expect(action.color).toEqual(color)
      })
    })

    describe('setError', () => {
      test('Should have the correct type', () => {
        const action = colorsFilter.setError()
        expect(action.type).toEqual(colorsFilter.SET_ERROR)
      })

      test('Should have the correct value', () => {
        const error = 'Error'
        const action = colorsFilter.setError(error)
        expect(action.error).toEqual(error)
      })
    })
  })

  describe('Reducer', () => {
    describe('setColors', () => {
      test('Should set colors', () => {
        const newState = colorsFilter.default(
          colorsFilter.defaultState,
          colorsFilter.setColors(MOCK_COLORS)
        )

        expect(newState.colors).toEqual(MOCK_COLORS)
      })
    })

    describe('selectColor', () => {
      test('Should set the `selectedColor` property', () => {
        const color = MOCK_COLORS[0]
        const newState = colorsFilter.default(
          colorsFilter.defaultState,
          colorsFilter.selectColor(color)
        )

        expect(newState.selectedColor).toEqual(color)
      })
    })

    describe('setError', () => {
      test('Should set the `error` property', () => {
        const error = 'Error'
        const newState = colorsFilter.default(
          colorsFilter.defaultState,
          colorsFilter.setError(error)
        )

        expect(newState.error).toEqual(error)
      })
    })
  })
})
