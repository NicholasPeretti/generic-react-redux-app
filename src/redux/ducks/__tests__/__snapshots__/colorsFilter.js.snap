// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colors Filter Reducer setColors Should set colors and colorsMap 1`] = `
Immutable.Map {
  "red": "red",
  "blue": "blue",
  "green": "green",
  "black": "black",
  "yellow": "yellow",
  "white": "white",
  "silver": "silver",
}
`;

exports[`Colors Filter defaultState should match snapshot 1`] = `
Object {
  "colors": Array [],
  "colorsMap": Immutable.Map {},
  "selectedColor": null,
}
`;
